network/gossip/libp2p/sporking_test.go

```diff
@@ -118,18 +107,127 @@ func (h *SporkingTestSuite) TestProtocolChangeAfterHardSpoon() {
+	node2key, err := generateNetworkingKey("def")
+	assert.NoError(h.T(), err)
+	node2, addr2 := h.CreateNode("node1", node2key, "0.0.0.0", "0", rootIDBeforeSpork, nil, false)
+	defer h.StopNode(node1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// spork topic is derived by suffixing the channelID with the root block ID
+	topicBeforeSpork := testChannelID + "-" + rootIDBeforeSpork
+
+	// both nodes are initially on the same spork and subscribed to the same topic
+	_, err = node1.Subscribe(ctx, topicBeforeSpork)
+	assert.NoError(h.T(), err)
+	sub2, err := node2.Subscribe(ctx, topicBeforeSpork)
+	assert.NoError(h.T(), err)
+
+	// add node 2 as a peer of node 1
+	err = node1.AddPeer(ctx, addr2)
+	assert.NoError(h.T(), err)
+
+	// let the node form the mesh
+	time.Sleep(time.Second)
+
+	// assert that node 1 can successfully send a message to node 2 via PubSub
+	testOneToKMessagingSucceeds(ctx, h.T(), node1, sub2, topicBeforeSpork)
+
+	// mimic that node1 now is part of the new spork while node2 remains on the old spork
+
+	// new root id after spork
+	rootIDAfterSpork := unittest.BlockFixture().ID().String()
+
+	// topic after the spork
+	topicAfterSpork := testChannelID + "-" + rootIDAfterSpork
+
+	// unsubscribe node1 from topicBeforeSpork and subscribe it to topicAfterSpork
+	err = node1.UnSubscribe(topicBeforeSpork)
+	assert.NoError(h.T(), err)
+	_, err = node1.Subscribe(ctx, topicAfterSpork)
+	assert.NoError(h.T(), err)
+
+	// assert that node 1 can no longer send a message to node 2 via PubSub
+	testOneToKMessagingFails(ctx, h.T(), node1, sub2, topicAfterSpork)
+}
+
+func testOneToOneMessagingSucceeds(t *testing.T, sourceNode *P2PNode, dstnAddress NodeAddress) {
+	// create stream from node 1 to node 2
+	s, err := sourceNode.CreateStream(context.Background(), dstnAddress)
+	// assert that stream creation succeeded
+	assert.NoError(t, err)
+	assert.NotNil(t, s)
+}
+
+func testOneToOneMessagingFails(t *testing.T, sourceNode *P2PNode, dstnAddress NodeAddress) {
+	// create stream from source node to destination address
+	_, err := sourceNode.CreateStream(context.Background(), dstnAddress)
 	// assert that stream creation failed
-	assert.Error(h.T(), err)
+	assert.Error(t, err)
 	// assert that it failed with the expected error
-	assert.Regexp(h.T(), ".*protocol not supported.*", err)
+	assert.Regexp(t, ".*failed to negotiate security protocol.*|.*protocol not supported.*", err)
+}
+
+func testOneToKMessagingSucceeds(ctx context.Context,
+	t *testing.T,
+	sourceNode *P2PNode,
+	dstnSub *pubsub.Subscription,
+	topic string) {
+
+	// send a 1-k message from source node to destination node
+	payload := []byte("hello")
+	err := sourceNode.Publish(ctx, topic, payload)
+	assert.NoError(t, err)
+
+	// assert that the message is received by the destination node
+	unittest.AssertReturnsBefore(t, func() {
+		msg, err := dstnSub.Next(ctx)
+		assert.NoError(t, err)
+		assert.Equal(t, payload, msg.Data)
+	}, 5*time.Second)
```

Comment:
We have 5 seconds here to check a successful message but only 1 second to check a failed message. If it takes say 2 seconds to send a 1-k message, `testOneToKMessagingFails` will pass even if the message does go through. 

Do you know how long the message publish takes in practice?

Reply:
In practice is should take way under a second once the mesh is formed. I agree, the timeout for the failed message should also be large (atleast 1.5 second). 
Lemme fine tune those numbers a little